src/utils/authentication.utils.ts

```diff
@@ -0,0 +1,60 @@
+import { Token, AccessToken } from 'devu-shared-modules'
+
+import RequestService from 'services/request.service'
+
+import store from 'redux/store'
+import { RootState } from 'redux/reducers'
+import { SET_USER } from 'redux/types/user.types'
+
+export function decodeAccessToken(accessToken: string): AccessToken {
+  const base64Url = accessToken.split('.')[1]
+  const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/')
+  const jsonPayload = decodeURIComponent(
+    atob(base64)
+      .split('')
+      .map((c) => '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2))
+      .join(''),
+  )
+
+  return JSON.parse(jsonPayload)
+}
+
+function timeInSecondsToEpoch(): number {
+  return Math.round(Date.now() / 1000)
+}
+
+function setToken({ accessToken }: Token) {
+  const state: RootState = store.getState()
+  const payload = { ...state.user, accessToken }
+
+  store.dispatch({ type: SET_USER, payload: payload })
+
+  return accessToken
+}
+
+/**
+ * Gets your accessToken from the state
+ * Using this function ensures that your accessToken is always up to date
+ * (aka it refreshes it for you if it's expired or nearing expiration)
+ */
+export function getToken(): Promise<string> | string {
+  const state: RootState = store.getState()
+  const { accessToken } = state.user
+
+  if (!accessToken) return ''
+
+  const decodedToken = decodeAccessToken(accessToken)
+  const expirationCutoff = timeInSecondsToEpoch() - 60 // within 60 seconds of expiration
+
+  if (decodedToken.exp && decodedToken.exp < expirationCutoff) return accessToken
+
+  return RequestService.get('/api/login/refresh', { credentials: 'include' })
+    .then(setToken)
+    .catch(() => {
+      // What do we want to do if a user attempts to update and their refreshToken is no longer good?
+      // Just reload the page for now?
```

Comment:
In this case they should go through the auth flow again - the easiest way to accomplish that is likely to just reload and have the normal auth flow take over. I know we discussed having them re-auth within X hours of their refresh token expiring, though the issue there is we have no way of knowing the expiry time of the refresh token as its httpOnly - will have to revisit and think about that a bit to hopefully make the UX not terrible.

Reply:
This comment was largely old, I updated it. Since we updated the auth flow to better handle this, I'm not totally worried about this.